### DISCUSSION TITLE
What are the prerequisites to learning Truffle?

### DISCUSSION THREAD

**Discussion by jeshan** [oracle/graal#2890] [Q&A] [answered]
**Describe the issue**
I'm looking to learn the Truffle api but I still have close to zero clue about language implementation.
When I read blogs like this:
https://medium.com/graalvm/graalvm-ten-things-12d9111f307d

I see the author @chrisseaton writes:
> Anyone who’s done just an undergraduate course in programming language implementation should have the basic skills needed.

As somebody who didn't study compiler design, I'd like to ask: What would those skills be?
Are there specific chapters in the ["Dragon book"](https://en.wikipedia.org/wiki/Compilers:_Principles,_Techniques,_and_Tools) that we should be reading? Would you recommend such a book in the first place?

**Comment by chrisseaton** [comment, marked as answer]
> Are there specific chapters in the "Dragon book" that we should be reading? Would you recommend such a book in the first place?

I wouldn't recommend the Aho et al book for writing Truffle languages. It's very old-fashioned and pretty much the opposite of the way Truffle does things.

I'd recommend (in order):

* [Writing An Interpreter In Go](https://interpreterbook.com/)
* [Writing A Compiler In Go](https://compilerbook.com/)
* [Crafting Interepters](https://craftinginterpreters.com/)

Then come back and look at Truffle. Maybe we should write a book based on people having that foundation.

**Comment by jeshan** [comment]
Thanks Chris, that's good enough for me.

**Comment by chumer** [comment]
Yes we should write an extensive manual for people with that background. 

In the meantime you can start learning Truffle here:
https://www.graalvm.org/graalvm-as-a-platform/language-implementation-framework/
